src/main/java/com/squareup/javapoet/TypeName.java

```diff
@@ -163,11 +163,12 @@ public TypeName unbox() {
     if (this == o) return true;
     if (o == null) return false;
     if (getClass() != o.getClass()) return false;
-    return toString().equals(o.toString());
+    return keyword.equals(((TypeName) o).keyword);
   }
 
   @Override public int hashCode() {
-    return toString().hashCode();
+    if (keyword == null) throw new AssertionError();
```

Comment:
See `emit(CodeWriter)` below, just mimicked that behaviour. Like `emit`, all subclasses of `TypeName` override `hashCode`, should work well with collections.